core/optaplanner-core-impl/src/main/java/org/optaplanner/core/api/solver/change/ProblemChangeDirector.java

```diff
@@ -113,4 +113,14 @@ <EntityOrProblemFact> void changeProblemProperty(EntityOrProblemFact problemFact
      * @param <EntityOrProblemFact> the object type
      */
     <EntityOrProblemFact> Optional<EntityOrProblemFact> lookUpWorkingObject(EntityOrProblemFact externalObject);
+
+    /**
+     * Calls variable listeners on the external changes submitted so far.
+     *
+     * @apiNote This happens automatically after the entire {@link ProblemChange} has been processed,
+     *          but this method allows the user to specifically request it in the middle of the {@link ProblemChange}.
+     * @implNote Calling this method may be expensive.
+     */
+    void triggerVariableListeners();
```

Comment:
Is it safe to call this method in a `ProblemChange`?
I think so, but a second pair of eyes always helps.

Reply:
The `ProblemChange` is executed on the `Solver` thread; from that perspective, it is safe. If it's called in the middle of the `ProblemChange` when the model is not consistent, it's still safe, because we call the method after the `ProblemChange` anyway.
Only if someone uses a stateful shadow variable listener, it may cause trouble because of duplicate processing.

Reply:
Is there duplicate processing?
Once the listener is processed once, it gets removed from the queue, never to be processed again. The variable would have to change again for it to show up in the queue again.